cpp/src/arrow/compute/kernels/vector_swizzle_test.cc

```diff
@@ -0,0 +1,1065 @@
+    int64_t max_index, const std::string& expected_str, bool validity_must_be_null) {
+  for (const auto& output_type : kIntegerTypes) {
+    ARROW_SCOPED_TRACE("Output type: " + output_type->ToString());
+    auto expected = ArrayFromJSON(output_type, expected_str);
+    DoTestInversePermutationForInputTypes(input_types, std::forward<InputFunc>(input),
+                                          max_index, output_type, expected,
+                                          validity_must_be_null);
+  }
+}
+
+void TestInversePermutationForInputOutputTypes(
+    const std::vector<std::shared_ptr<DataType>>& input_types,
+    const std::vector<std::shared_ptr<DataType>>& output_types,
+    const std::string& indices_str, const std::vector<std::string>& indices_chunked_str,
+    int64_t max_index, const std::string& expected_str, bool validity_must_be_null) {
+  {
+    ARROW_SCOPED_TRACE("Array");
+    DoTestInversePermutationForInputOutputTypes(
+        input_types, output_types,
+        [&](const std::shared_ptr<DataType>& input_type) {
+          return ArrayFromJSON(input_type, indices_str);
+        },
+        max_index, expected_str, validity_must_be_null);
+  }
+  {
+    ARROW_SCOPED_TRACE("Chunked");
+    DoTestInversePermutationForInputOutputTypes(
+        input_types, output_types,
+        [&](const std::shared_ptr<DataType>& input_type) {
+          return ChunkedArrayFromJSON(input_type, indices_chunked_str);
+        },
+        max_index, expected_str, validity_must_be_null);
+  }
+}
+
+void TestInversePermutationSigned(const std::string& indices_str,
+                                  const std::vector<std::string>& indices_chunked_str,
+                                  int64_t max_index, const std::string& expected_str,
+                                  bool validity_must_be_null = false) {
+  TestInversePermutationForInputOutputTypes(kSignedIntegerTypes, kIntegerTypes,
+                                            indices_str, indices_chunked_str, max_index,
+                                            expected_str, validity_must_be_null);
+}
+
+void TestInversePermutation(const std::string& indices_str,
+                            const std::vector<std::string>& indices_chunked_str,
+                            int64_t max_index, const std::string& expected_str,
+                            bool validity_must_be_null = false) {
+  TestInversePermutationForInputOutputTypes(kIntegerTypes, kIntegerTypes, indices_str,
+                                            indices_chunked_str, max_index, expected_str,
+                                            validity_must_be_null);
+}
+
+}  // namespace
+
+TEST(InversePermutation, InvalidOutputType) {
+  {
+    ARROW_SCOPED_TRACE("Output type float");
+    auto indices = ArrayFromJSON(int32(), "[]");
+    ASSERT_RAISES_WITH_MESSAGE(
+        Invalid, "Invalid: Output type of inverse_permutation must be integer, got float",
```

Comment:
This could usefully be a `TypeError`, allowing it to distinguish with errors in the data.

Reply:
Done.